### PR TITLE
Add validator for navigation in backend

### LINKED ISSUE #4491: Hide plugins in the top navigation that only contain sidemenu
I feel like I'm having one of those days where errors are following me about.

I installed two plugins from the marketplace.

1st plugin (main plugin) can be found here: https://octobercms.com/plugin/jorgeandrade-subscribe

2nd plugin (plugin extras) can be found here: https://octobercms.com/plugin/jorgeandrade-subscribeplus

When installing both the plugins I get the following error in the top navigation, see screenshot:

nobe-end

Looking at their authors code we have the following:

Plugin 1:

```php
    public function registerNavigation()
    {
        return [
            'subscribe' => [
                'label' => 'Subscribers',
                'url' => \Backend::url('jorgeandrade/subscribe/subscribers'),
                'icon' => 'icon-rss',
                'permissions' => ['jorgeandrade.subscribe.*'],
                'order' => 500,

                'sideMenu' => [
                    'subscribers' => [
                        'label' => 'Subscribers',
                        'icon' => 'icon-rss',
                        'url' => \Backend::url('jorgeandrade/subscribe/subscribers'),
                        'permissions' => ['jorgeandrade.subscribe.access_subscribers'],
                    ],
                ],

            ],
        ];
    }
```

Plugin 2:

```php
    public function registerNavigation()
    {
        return [
            'subscribeplus' => [
                'sideMenu' => [
                    'subscribers' => [
                        'label'       => 'Subscribers',
                        'icon'        => 'icon-rss',
                        'url'         => \Backend::url('jorgeandrade/subscribe/subscribers'),
                        'permissions' => ['jorgeandrade.subscribe.access_subscribers'],
                    ],
                    'lists' => [
                        'label'       => 'Subscribers Lists',
                        'icon'        => 'icon-list',
                        'url'         => \Backend::url('jorgeandrade/subscribeplus/lists'),
                        'permissions' => ['jorgeandrade.subscribeplus.lists'],
                    ],
                    'campaigns' => [
                        'label'       => 'Campaigns',
                        'icon'        => 'icon-newspaper-o',
                        'url'         => \Backend::url('jorgeandrade/subscribeplus/campaigns'),
                        'permissions' => ['jorgeandrade.subscribeplus.campaigns'],
                    ],
                    'templates' => [
                        'label'       => 'Templates',
                        'icon'        => 'icon-file-code-o',
                        'url'         => \Backend::url('jorgeandrade/subscribeplus/templates'),
                        'permissions' => ['jorgeandrade.subscribeplus.templates'],
                    ]
                ]

            ]
        ];
    }
```

Ripping out the HTML code from October we have this:

```html
<li class=" svg-icon-container svg-active-effects" role="listitem">
    <a href="">
        <span class="nav-icon">
            <i class=" "></i>
        </span>
        <span class="nav-label">
        </span>
    </a>
    <span class="counter empty" data-menu-id="subscribeplus">
    </span>
</li>
```

We can see the issue is with the `<a href="">`

This got me thinking and I came up with this code:

```html
<script>
$(function() {
    $('.mainmenu-nav a').each(function(){    
        if (!$(this).attr('href')) {
            $(this).parent().css("display", "none");
        }
    });
});
</script>
```

The above javascript will hide any blank plugins in the top navigation menu, helping to keep the top navigation cleaner.

### PR DESCRIPTION
**Pull request by Samuell1**
Resolves #4491

Tested with RainLab.Blog plugin

Main navigation:
obrázok

Nested sideMenu:
obrázok

### CODE REVIEW

**Comment by LukeTowers** [comment]
@bennothommo What do you think about wrapping this in a check to debug mode? If debug mode is enabled, throw an exception. If not, then log an error. I don't want to suddenly leave non-technical users with broken sites just because they have crap plugins (note, I know lots of people have prod sites with debug enabled but I'm fine with telling them to stop having debug enabled on production when they ask for help with this error).

File: modules/backend/classes/NavigationManager.php [hunk 1]
```diff
@@ -195,6 +197,19 @@ public function registerMenuItems($owner, array $definitions)
...
+            '*.label' => 'required',
+            '*.icon' => 'required',
+            '*.url' => 'required',
+            '*.sideMenu.*.label' => 'nullable|required',
+            '*.sideMenu.*.icon' => 'nullable|required',
+            '*.sideMenu.*.url' => 'nullable|required',
+        ]);
+
+        if ($validator->fails()) {
+            throw new SystemException('Error parsing navigation menu in plugin ' . $owner . ' (' . $validator->errors()->first() . ')');
```
Comment (LukeTowers) [inline modules/backend/classes/NavigationManager.php]:
```suggestion
            throw new SystemException('Invalid menu item detected in ' . $owner . '. Contact the plugin author to fix (' . $validator->errors()->first() . ')');
```

**Comment by bennothommo** [comment]
@LukeTowers sounds good to me. @Samuell1 would you mind adjusting the check based on debug mode as per Luke's [comments](https://github.com/octobercms/october/pull/4497#issuecomment-516190854)?

**Comment by w20k** [comment]
LGTM! Nice ;)

**Comment by ayumihamsaki** [comment]
@Samuell1 Thanks 👍

**Comment by bennothommo** [comment]
@Samuell1 I've tested this and it works, except for one thing: `icon` in the main nav and sub-nav items can actually be replaced with `iconSvg` and still work. If a plugin developer only uses `iconSvg`, the exception / log is thrown as `icon` is missing. You may need to adjust the validation to look for either of those values. 

I realise it's not specified in the comment, so apologies for that.

File: modules/backend/classes/NavigationManager.php [hunk 2]
```diff
@@ -195,6 +199,26 @@ public function registerMenuItems($owner, array $definitions)
             $this->items = [];
         }
 
+        $validator = Validator::make($definitions, [
+            '*.label' => 'required',
+            '*.icon' => 'required_without:*.iconSvg',
+            '*.iconSvg' => 'required_without:*.icon',
```
Comment (bennothommo) [inline modules/backend/classes/NavigationManager.php]: @Samuell1 You can drop this validation rule - the `*.icon` rule covers both cases.

File: modules/backend/classes/NavigationManager.php [hunk 3]
```diff
@@ -195,6 +199,26 @@ public function registerMenuItems($owner, array $definitions)
...
+            '*.url' => 'required',
+            '*.sideMenu.*.label' => 'nullable|required',
+            '*.sideMenu.*.icon' => 'nullable|required_without:*.sideMenu.*.iconSvg',
+            '*.sideMenu.*.iconSvg' => 'nullable|required_without:*.sideMenu.*.icon',
```
Comment (bennothommo) [inline modules/backend/classes/NavigationManager.php]: @Samuell1 This one too :)

**Comment by bennothommo** [comment]
Blocked until development begins on build 459.

**Pull request merged**